### PR TITLE
Rework Scale System

### PR DESCRIPTION
**Pull request by caseycrogers**
## :scroll: Description

This PR completely reworks the `feedback_widget` to use `CustomMultiChildLayout` instead of `Stack`.

## :bulb: Motivation and Context

This has a number of advantages:

1. The layout of each of the three key widgets (controls, sheet, screenshot) relative to the others is explicit and reactive. Previously, depending on the user device and orientation, the widgets could end up overlapping and/or looking awkward.
2. It becomes much easier to reason about and modify the widgets' layout.
3. All the elements animate in sync with each other.

I was especially finding the existing system to be too unmanageable while trying to update the bottom sheet to be a draggable sheet.

TODO:

1. Don't bug out when the keyboard is shown ✔️
2. Tweak the layout and incoming animation ✔️
3. Make the controls column wrap if it is too long for the screen (particularly relevant in landscape mode) ✔️
4. Expose `sheetFraction` in `FeedbackThemeData`? ❔ 

## :green_heart: How did you test it?

Goofed off with the example widget. Soon to add this to my existing app via a path dependency.

## :pencil: Checklist

- [ ] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] I updated the docs if needed
- [ ] All tests passing
- [ ] No breaking changes

### CODE REVIEW

File: feedback/lib/src/feedback_widget.dart [hunk 1]
```diff
@@ -294,3 +306,79 @@ class FeedbackWidgetState extends State<FeedbackWidget>
...
+class _FeedbackLayoutDelegate extends MultiChildLayoutDelegate {
+  _FeedbackLayoutDelegate({
+    required this.query,
+    required this.sheetFraction,
+    required this.progress,
+  });
+
+  MediaQueryData query;
+  double sheetFraction;
+  double progress;
```
Comment (ueman) [inline feedback/lib/src/feedback_widget.dart:323]: This is the animation progress, right?

File: feedback/lib/src/feedback_widget.dart [hunk 2]
```diff
@@ -294,3 +306,79 @@ class FeedbackWidgetState extends State<FeedbackWidget>
...
+
+class _FeedbackLayoutDelegate extends MultiChildLayoutDelegate {
+  _FeedbackLayoutDelegate({
+    required this.query,
+    required this.sheetFraction,
+    required this.progress,
+  });
+
+  MediaQueryData query;
+  double sheetFraction;
```
Comment (ueman) [inline feedback/lib/src/feedback_widget.dart:322]: This is how much of the vertical screen the bottom sheet covers, right?

File: feedback/lib/src/feedback_widget.dart [hunk 3]
```diff
@@ -294,3 +306,79 @@ class FeedbackWidgetState extends State<FeedbackWidget>
     FocusScope.of(context).requestFocus(FocusNode());
   }
 }
+
+const _screenshotId = 'screenshot_id';
+const _controlsColumnId = 'controls_column_id';
+const _sheetId = 'sheet_id';
+
+class _FeedbackLayoutDelegate extends MultiChildLayoutDelegate {
```
Comment (ueman) [inline feedback/lib/src/feedback_widget.dart:314]: Do you mind explaining why this is needed? I'm not really familiar with `CustomMultiChildLayout` and `MultiChildLayoutDelegate`.

**Review by ueman — COMMENTED** [review]
Hey, thank you very much for this. 

> Previously, depending on the user device and orientation, the widgets could end up overlapping and/or looking awkward.

Yeah, I agree. That's still from when I started with Flutter 😅

File: feedback/lib/src/feedback_widget.dart (hunk 1 above)
Comment (caseycrogers) [reply feedback/lib/src/feedback_widget.dart]: Yep, renamed it as such to make it more clear.

File: feedback/lib/src/feedback_widget.dart (hunk 2 above)
Comment (caseycrogers) [reply feedback/lib/src/feedback_widget.dart:323]: Yep.

File: feedback/lib/src/feedback_widget.dart (hunk 3 above)
Comment (caseycrogers) [reply feedback/lib/src/feedback_widget.dart:315]: `CustomMultiChildLayout` is a convenience widget over the underlying Flutter layout system (I think? I'm not actually an expert on it). So when you use a `Stack` or `Column` etc.,   it's using something pretty similar under the hood. It's a lot simpler than it looks at first glance.

It has an arbitrary number of children that are all passed in with an ID. The delegate interacts with the children using their IDs..
You call `layoutChild(someId, constraints)` to tell a given child widget to build itself within the given box constraints. This is like wrapping a `ConstrainedBox` around the child.
You call `positionChild(...)` to position it. This is like wrapping `Positioned` around it in a stack.
The key advantage is that you choose the order in which to lay the children out and `layoutChild` gives you the size the child ended up taking up. This way, children you layout later can have arbitrary dependencies on the sizes of children you laid out earlier.

The biggest reason I used it over just a stack is that the screenshot widget's size and position are dependent on the width of the controls column (if we want the layout to be properly reactive).
To achieve this inter-dependency with a stack, I'd either have to (AFAIK):

1. Use `GlobalContext` and a post frame callback to get the width after the first draw
2. Force the controls column to be a preset width

Both felt sloppy and are known anti patterns.

Comment (ueman) [reply feedback/lib/src/feedback_widget.dart:315]: Thanks!

**Comment by caseycrogers** [comment]
@ueman The PR is mostly done. I've also made the controls column wrap if it's too tall to fit.

I have a couple question around the sheet size if you have a minute:

1. What should the default height of the sheet be? I've picked 25% of the screen because it was close to the original.
2. What should we do about the sheet in landscape mode? The screenshot is pretty small, but the sheet is also unusably small. Rock and hard place.
3. Should we let developers customize the sheet's height via `FeedbackTheme`? Should the parameter be a fraction of screen size, or number of pixels?

**Comment by ueman** [comment]
> I've also made the controls column wrap if it's too tall to fit.

Awesome.

> 1. What should the default height of the sheet be? I've picked 25% of the screen because it was close to the original.

That's fine for me.

> 2. What should we do about the sheet in landscape mode? The screenshot is pretty small, but the sheet is also unusably small. Rock and hard place.

Maybe in landscape it shouldn't be a bottom sheet but layout where both things are next to each other? That's probably more sensible because we're having much more horizontal space than vertical. That's probably work for another PR, though. For now, leaving it as it is fine as long as it's not worse than before.

> 3. Should we let developers customize the sheet's height via `FeedbackTheme`? Should the parameter be a fraction of screen size, or number of pixels? Both and the framework will pick the min or max between the two?

I think I prefer having a fraction available in the FeedbackTheme. It seems reasonable. The docs for it should probably say that the recommended height of the sheet is around a quarter to a third.

If you're having trouble with the golden images again, just ignore them and I'll fix it them you want.

**Comment by caseycrogers** [comment]
All done! I left the golden images and changelog/version bumping to you.

The changes to support a draggable bottom sheet are near done too so if you want to do one update, I could get the bottom sheet done pretty quick once this is merged.

Let me know if you think this PR needs any changes!

**Comment by ueman** [comment]
I'm going to close this, as the changes are also included in #149

**Pull request closed without merging**